File: refactor_sizes.py
import re
import os

files_to_process = [
    'd:/habitz/lib/Features/solo/widgets/challenge_result_card.dart',
    'd:/habitz/lib/Features/solo/widgets/smart_feed_card.dart'
]

import_statement = "import '../constants/solo_constants.dart';"

for file_path in files_to_process:
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()

    if import_statement not in content:
        content = re.sub(r"(import 'package:flutter/material.dart';)", r"\1\n" + import_statement, content)

    # Use a set of rules to replace sizes
    # fontSize: 18 -> fontSize: 18.rs(context)
    content = re.sub(r'(fontSize:\s*)([0-9]+(?:\.[0-9]+)?)(?![\.r])', r'\1\2.rs(context)', content)

    # SizedBox(height: 16) -> SizedBox(height: 16.rs(context))
    content = re.sub(r'(SizedBox\s*\(\s*(?:width|height):\s*)([0-9]+(?:\.[0-9]+)?)(?![\.r])', r'\1\2.rs(context)', content)

    # padding: const EdgeInsets.all(22) -> padding: EdgeInsets.all(22.rs(context))
    content = re.sub(r'const\s*(EdgeInsets\.all\()([0-9]+(?:\.[0-9]+)?)\)', r'\1\2.rs(context))', content)
    
    # helper functions for symmetric, only, fromLTRB
    def res_sym(m):
        res = m.group(1).replace('const ', '')
        inner = m.group(2)
        inner = re.sub(r'(horizontal|vertical):\s*([0-9]+(?:\.[0-9]+)?)', r'\1: \2.rs(context)', inner)
        return res + inner + ')'
        
    content = re.sub(r'(const\s*EdgeInsets\.symmetric\()([^)]+)\)', res_sym, content)

    def res_only(m):
        res = m.group(1).replace('const ', '')
        inner = m.group(2)
        inner = re.sub(r'(left|right|top|bottom):\s*([0-9]+(?:\.[0-9]+)?)', r'\1: \2.rs(context)', inner)
        return res + inner + ')'

    content = re.sub(r'(const\s*EdgeInsets\.only\()([^)]+)\)', res_only, content)

    def res_fromLTRB(m):
        res = m.group(1).replace('const ', '')
        nums = m.group(2).split(',')
        new_nums = [n.strip() + '.rs(context)' for n in nums]
        return res + ', '.join(new_nums) + ')'

    content = re.sub(r'(const\s*EdgeInsets\.fromLTRB\()([^)]+)\)', res_fromLTRB, content)

    # Other common sizes
    # width: 52
    # height: 52
    content = re.sub(r'((?:^\s*|\b)(?:width|height|size|blurRadius):\s*)([0-9]+(?:\.[0-9]+)?)(?![\.r0-9a-zA-Z])', r'\1\2.rs(context)', content)
    
    # BorderRadius.circular(28) -> BorderRadius.circular(28.rs(context))
    content = re.sub(r'(BorderRadius\.circular\()([0-9]+(?:\.[0-9]+)?)\)', r'\1\2.rs(context))', content)

    # Remove duplicates like 18.rs(context).rs(context) which might happen
    content = re.sub(r'\.rs\(context\)\.rs\(context\)', '.rs(context)', content)

    with open(file_path, 'w', encoding='utf-8') as f:
        f.write(content)

print('Done')

File: refactor_sizes.dart
import 'dart:io';

void main() {
  final files = [
    'lib/Features/solo/widgets/challenge_result_card.dart'
  ];
  
  final importStatement = "import '../constants/solo_constants.dart';";

  for (final filePath in files) {
    final file = File(filePath);
    if (!file.existsSync()) {
      print('File not found: $filePath');
      continue;
    }
    
    var content = file.readAsStringSync();
    
    if (!content.contains(importStatement)) {
       content = content.replaceFirst("import 'package:flutter/material.dart';", "import 'package:flutter/material.dart';\n$importStatement");
    }

    // fontSize: 18 -> fontSize: 18.rs(context)
    content = content.replaceAllMapped(RegExp(r'(fontSize:\s*)([0-9]+(\.[0-9]+)?)(?![\.r])'), (match) {
      return '${match.group(1)}${match.group(2)}.rs(context)';
    });

    // SizedBox(height: 16) -> SizedBox(height: 16.rs(context))
    content = content.replaceAllMapped(RegExp(r'(SizedBox\s*\(\s*(?:width|height):\s*)([0-9]+(\.[0-9]+)?)(?![\.r])'), (match) {
      return '${match.group(1)}${match.group(2)}.rs(context)';
    });

    // padding: const EdgeInsets.all(22) -> padding: EdgeInsets.all(22.rs(context))
    content = content.replaceAllMapped(RegExp(r'const\s*(EdgeInsets\.all\()([0-9]+(\.[0-9]+)?)\)'), (match) {
      return '${match.group(1)}${match.group(2)}.rs(context))';
    });

    // EdgeInsets.symmetric
    content = content.replaceAllMapped(RegExp(r'(const\s+EdgeInsets\.symmetric\()([^)]+)\)'), (match) {
      var res = match.group(1)!.replaceAll('const ', '');
      var inner = match.group(2)!;
      inner = inner.replaceAllMapped(RegExp(r'(horizontal|vertical):\s*([0-9]+(\.[0-9]+)?)'), (m) => '${m.group(1)}: ${m.group(2)}.rs(context)');
      return res + inner + ')';
    });

    // EdgeInsets.only
    content = content.replaceAllMapped(RegExp(r'(const\s+EdgeInsets\.only\()([^)]+)\)'), (match) {
      var res = match.group(1)!.replaceAll('const ', '');
      var inner = match.group(2)!;
      inner = inner.replaceAllMapped(RegExp(r'(left|right|top|bottom):\s*([0-9]+(\.[0-9]+)?)'), (m) => '${m.group(1)}: ${m.group(2)}.rs(context)');
      return res + inner + ')';
    });
    
    // EdgeInsets.fromLTRB
    content = content.replaceAllMapped(RegExp(r'(const\s+EdgeInsets\.fromLTRB\()([^)]+)\)'), (match) {
      var res = match.group(1)!.replaceAll('const ', '');
      var numsString = match.group(2)!;
      var newNums = numsString.split(',').map((n) {
          final trimmed = n.trim();
          if (trimmed.isEmpty) return trimmed;
          if (RegExp(r'^[0-9]+(\.[0-9]+)?$').hasMatch(trimmed)) {
             return '$trimmed.rs(context)';
          }
          return trimmed;
      }).join(', ');
      return res + newNums + ')';
    });
    
    // widths, heights, size
    content = content.replaceAllMapped(RegExp(r'((?:^\s*|\s|\b)(?:width|height|size|blurRadius):\s*)([0-9]+(\.[0-9]+)?)(?![\.r0-9a-zA-Z])'), (match) {
       return '${match.group(1)}${match.group(2)}.rs(context)';
    });

    // BorderRadius.circular
    content = content.replaceAllMapped(RegExp(r'(BorderRadius\.circular\()([0-9]+(\.[0-9]+)?)\)'), (match) {
        return '${match.group(1)}${match.group(2)}.rs(context))';
    });

    file.writeAsStringSync(content);
  }
}
